pallets/stableswap/src/lib.rs

```diff
@@ -878,6 +880,113 @@ pub mod pallet {
+			);
+
+			let pool = Pools::<T>::get(pool_id).ok_or(Error::<T>::PoolNotFound)?;
+			let pool_account = Self::pool_account(pool_id);
+			let initial_reserves = pool
+				.reserves_with_decimals::<T>(&pool_account)
+				.ok_or(Error::<T>::UnknownDecimals)?;
+			let share_issuance = T::Currency::total_issuance(pool_id);
+
+			ensure!(
+				share_issuance == share_amount
+					|| share_issuance.saturating_sub(share_amount) >= T::MinPoolLiquidity::get(),
+				Error::<T>::InsufficientLiquidityRemaining
+			);
+
+			// We want to ensure that given min amounts are correct. It must contain all pool assets.
+			// We convert vec of min amounts to a map.
+			// We first ensure the length , and if any asset is not found later on, we can return an error.
+			ensure!(min_amounts_out.len() == pool.assets.len(), Error::<T>::IncorrectAssets);
+			let mut min_amounts_out: BTreeMap<T::AssetId, Balance> =
+				min_amounts_out.into_iter().map(|v| (v.asset_id, v.amount)).collect();
+
+			// Store the amount of each asset that is transferred. Used as info in the event.
+			let mut amounts = Vec::with_capacity(pool.assets.len());
+
+			// 1. Calculate amount of each asset
+			// 2. ensure min amount is respected
+			// 3. transfer amount to user
+			for asset_id in pool.assets.iter() {
+				ensure!(
+					Self::is_asset_allowed(pool_id, *asset_id, Tradability::REMOVE_LIQUIDITY),
+					Error::<T>::NotAllowed
+				);
+				let min_amount = min_amounts_out.remove(asset_id).ok_or(Error::<T>::IncorrectAssets)?;
+				let reserve = T::Currency::free_balance(*asset_id, &pool_account);
+
+				// Special case when withdrawing all remaining pool shares, so we can directly send all the remaining assets to the user.
+				let amount = if share_amount == share_issuance {
+					// no need to ensure the min amounts in this case
+					reserve
+				} else {
+					let amount =
+						hydra_dx_math::stableswap::calculate_liquidity_out(reserve, share_amount, share_issuance)
+							.ok_or(ArithmeticError::Overflow)?;
+					ensure!(amount >= min_amount, Error::<T>::SlippageLimit);
+					amount
+				};
+
+				T::Currency::transfer(*asset_id, &pool_account, &who, amount)?;
+				amounts.push(AssetAmount {
+					asset_id: *asset_id,
+					amount,
+				});
+			}
+
+			// Burn shares
+			T::Currency::withdraw(pool_id, &who, share_amount)?;
+
+			// All done and updated. let's call the on_liquidity_changed hook.
+			// TODO: liquidity change hook does not work when we are removing all liquidity.
+			// it is tricky to update the hooks as it works with share prices of each asset, which is not possible to calculate with total issuance == 0.
+			// we should consider to introduce new hook to inform that this pool is being removed.
+			if share_amount != share_issuance {
+				Self::call_on_liquidity_change_hook(pool_id, &initial_reserves, share_issuance)?;
+			} else {
+				Pools::<T>::remove(pool_id);
+				let _ = AssetTradability::<T>::clear_prefix(pool_id, MAX_ASSETS_IN_POOL, None);
```

Comment:
Why is this line needed? Anyway, it is not tested either, pls consider adding an assertion if possible.

Reply:
we destroy the pool and clear all related storages when all liquidity from this pool is removed.